File: gurukrupa_customizations/gurukrupa_customizations/report/monthly_in_out/monthly_in_out.py
# Copyright (c) 2023, 8848 Digital LLP and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from datetime import timedelta, datetime
from frappe.utils import flt, getdate, add_days, format_time, today, add_to_date, get_time
from gurukrupa_customizations.gurukrupa_customizations.doctype.manual_punch.manual_punch import get_checkins

STATUS = {
	"Absent" : "A",
	"Present" : "P",
	"Half Day" : "HD",
	"Paid Leave" : "PL",
	"Casual Leave" : "CL",
	"Sick Leave" : "SL",
	"Leave Without Pay" : "LWP",
	"Outdoor Duty" : "OD",
	"Maternity Leave" : "ML",
}

def execute(filters=None):
	columns = get_columns(filters)
	data = get_data(filters)
	return columns, data

def get_data(filters=None):
	conditions = get_conditions(filters)
	data = frappe.db.sql(f"""select 
			at.attendance_date,concat(TIME_FORMAT(st.start_time,"%H:%i:%s"), " TO ", TIME_FORMAT(st.end_time,"%H:%i:%s")) as shift, time(at.in_time) as in_time, time(at.out_time) as out_time, 
			sec_to_time(at.working_hours*3600) as spent_hours, at.late_entry, if(at.late_entry,timediff(time(at.in_time), st.start_time),Null) as late_hrs,
			if(at.early_exit,timediff(st.end_time,time(at.out_time)),Null) as early_hrs, pol.hrs as p_out_hrs, 
			sec_to_time(
				if((at.attendance_request is not null or (at.status = "On Leave" and at.leave_type in (select name from `tabLeave Type` where is_lwp = 0))),
					st.shift_hours,
					at.working_hours)
				*3600
					+ if(at.late_entry=0 and time(at.in_time) > time(st.start_time),
							time_to_sec(timediff(time(at.in_time), st.start_time)), 0)
					- if(time(at.in_time) < time(st.start_time),
							time_to_sec(timediff(st.start_time, time(at.in_time))), 0)
					- if(at.out_time > timestamp(date(at.in_time),st.end_time),
							time_to_sec(timediff(at.out_time, timestamp(date(at.in_time),st.end_time))), 0)
					- ifnull(time_to_sec(pol.hrs),0)
					+ (select ifnull(sum(time_to_sec(pl.total_hours)),0) from `tabPersonal Out Log` pl 
						where pl.is_cancelled = 0 and pl.employee = at.employee and pl.date = at.attendance_date and pl.out_time >= st.end_time)
				) as net_wrk_hrs,
			st.shift_hours, 
			if(st.working_hours_threshold_for_half_day>at.working_hours and at.working_hours > 0,1,0) as lh,
			ot.allowed_ot as ot_hours, 
			ifnull(at.leave_type, at.status) as status, at.attendance_request
			from 
			`tabAttendance` at left join 
			`tabEmployee` emp on at.employee = emp.name left join
			`tabShift Type` st on emp.default_shift=st.name left join
			(select employee, date, sec_to_time(sum(time_to_sec(total_hours))) as hrs from 
			`tabPersonal Out Log` where is_cancelled = 0 group by employee,date) pol on at.attendance_date = pol.date and at.employee = pol.employee left join
			(select * from `tabOT Log` where is_cancelled=0) ot on at.attendance_date = ot.attendance_date and at.employee = ot.employee

			where at.docstatus = 1 {conditions} 
			order by at.attendance_date asc""", as_dict=1)
	
	if not data:
		return
	
	data = process_data(data, filters)
	totals = get_totals(data, filters.get("employee"))
	data += totals
	return data

def get_totals(data, employee):	
	totals = {
		"status": "Total Hours",
		"net_wrk_hrs": timedelta(0),
		"spent_hours": timedelta(0),
		"late_hrs": timedelta(0),
		"early_hrs": timedelta(0),
		"p_out_hrs": timedelta(0),
		"net_wrk_hrs": timedelta(0),
		"ot_hours": timedelta(0),
		"total_pay_hrs": timedelta(0),
	}
	late_count = 0
	penalty_days = 0
	for row in data:
		totals["net_wrk_hrs"] += (row.get("net_wrk_hrs") or timedelta(0))
		totals["total_pay_hrs"] += (row.get("total_pay_hrs") or timedelta(0))
		totals["ot_hours"] += (row.get("ot_hours") or timedelta(0))
		totals["early_hrs"] += (row.get("early_hrs") or timedelta(0))
		totals["late_hrs"] += (row.get("late_hrs") or timedelta(0))
		totals["p_out_hrs"] += (row.get("p_out_hrs") or timedelta(0))
		totals["spent_hours"] += (row.get("spent_hours") or timedelta(0))
		if row.get("late_entry"):
			late_count += 1
		if row.get("shift_hours"):
			totals["shift_hours"] = row.get("shift_hours")
	
	if late_count > 4 and late_count < 10:
		penalty_days = 0.5
	if late_count >= 10 and late_count < 15:
		penalty_days = 1
	if late_count > 15:
		penalty_days = 1.5

	total_days = {"status":"Total Days"}
	conversion_factor = 3600 * flt(totals["shift_hours"])

	penalty_hrs = timedelta(hours=flt(totals["shift_hours"])*penalty_days)
	for key,value in totals.items():
		if key in ["status","shift_hours"]:
			continue
		total_days[key] = flt(value.total_seconds() / conversion_factor, 2)

	refund = {
		"ot_hours": "Refund Min(P.Hrs)",
		"total_pay_hrs" : min(frappe.db.get_value("Employee",employee,'allowed_personal_hours') or timedelta(0), (totals["early_hrs"]+totals["late_hrs"]+totals["p_out_hrs"]))
	}

	penalty_for_late_entry = {
		"ot_hours": "Penalty in Days",
		"total_pay_hrs" : penalty_days
	}

	net_pay_hrs = {
		"ot_hours": "Net Hrs",
		"total_pay_hrs" : totals["total_pay_hrs"] + refund["total_pay_hrs"] - penalty_hrs
	}

	net_pay_days = {
		"ot_hours": "Net Days",
		"total_pay_hrs" : flt(net_pay_hrs['total_pay_hrs'].total_seconds() / conversion_factor, 2)
	}

	net_pay_hrs_wo_ot = {
		"ot_hours": "Net Hrs w/o OT",
		"total_pay_hrs" : totals["total_pay_hrs"] + refund["total_pay_hrs"] - penalty_hrs - totals["ot_hours"]
	}

	net_pay_days_wo_ot = {
		"ot_hours": "Net Days w/o OT",
		"total_pay_hrs" : flt(net_pay_hrs_wo_ot['total_pay_hrs'].total_seconds() / conversion_factor, 2)
	}

	return [totals, total_days, refund, penalty_for_late_entry, net_pay_hrs, net_pay_days, net_pay_hrs_wo_ot, net_pay_days_wo_ot]

def process_data(data, filters):
	employee = filters.get("employee")
	from_date = filters.get("from_date")
	to_date = filters.get("to_date")
	processed = {}
	result = []
	holidays = []
	wo = []
	emp_det = frappe.db.get_value("Employee", employee, ["default_shift","holiday_list","date_of_joining"], as_dict=1)
	shift = emp_det.get("default_shift")
	shift_det = frappe.db.get_value("Shift Type", shift, ['shift_hours','holiday_list','start_time', 'end_time'], as_dict=1)
	shift_hours = flt(shift_det.get("shift_hours"))
	shift_name = f"{format_time(shift_det.get('start_time'))} To {format_time(shift_det.get('end_time'))}"
	checkins = frappe.db.sql(f"""select date(time) as login_date, attendance, count(name) as cnt from `tabEmployee Checkin` 
			  where time between '{from_date}' and '{add_days(to_date,1)}' and employee = '{employee}' group by attendance""", as_dict=1)
	checkins = {row.login_date: row.cnt for row in checkins}
	od = frappe.get_list("Employee Checkin",{'employee':employee,'source':"Outdoor Duty", "time": ['between',[from_date,add_days(to_date,1)]]},'date(time) as login_date', pluck='login_date',group_by='login_date')
	if shift and not emp_det.get('holiday_list'):
			emp_det['holiday_list'] = shift_det.get("holiday_list")
	
	if hl_name:=emp_det.get('holiday_list'):
		holidays = frappe.get_list("Holiday", {"parent": hl_name,
					"holiday_date":["between",[from_date, to_date]]}, ["holiday_date","weekly_off"], ignore_permissions=1)
		wo = [row.holiday_date for row in holidays if row.weekly_off]
		holidays = [row.holiday_date for row in holidays if not row.weekly_off]

	for row in data:
		if row.lh:
			row.status = 'LH'
		shift_hours_in_sec = row.shift_hours*3600
		if row.net_wrk_hrs.total_seconds() > shift_hours_in_sec or (shift_hours_in_sec - row.net_wrk_hrs.total_seconds()) < 60:
			row.net_wrk_hrs = timedelta(hours=row.shift_hours)
		row["total_pay_hrs"] = row.net_wrk_hrs + (row.get("ot_hours") or timedelta(0))
		row.status = STATUS.get(row.status) or row.status
		processed[row.attendance_date] = row

	ot_for_wo = frappe.get_all("OT Log", {"employee":employee,"attendance_date": ["between",[from_date,to_date]], "is_cancelled":0}, ["attendance_date","allowed_ot as ot_hours", "first_in as in_time", "last_out as out_time"])
	ot_for_wo = {row.attendance_date: row for row in ot_for_wo}
	date_range = get_date_range(from_date, to_date)
	for date in date_range:
		row = processed.get(date,ot_for_wo.get(date,{}))
		if date in od:
			row["status"] = "OD"
			if ot_hours:=row.get("ot_hours"):
				row['total_pay_hrs'] = ot_hours
		elif date in wo and (date >= getdate(emp_det.get("date_of_joining"))):
			status = "WO"
			date_time = datetime.combine(getdate(date), get_time(shift_det.start_time))
			if first_in_last_out := get_checkins(employee,date_time):		
				row["in_time"] = get_time(first_in_last_out[0].get("time"))
				row["out_time"] = get_time(first_in_last_out[-1].get("time"))
			if ot_hours:=row.get("ot_hours"):
				row['total_pay_hrs'] = ot_hours
		elif (date in holidays) and (date >= getdate(emp_det.get("date_of_joining"))):
			status = "H"
			row['net_wrk_hrs'] = timedelta(hours=shift_hours)
			row['total_pay_hrs'] = timedelta(hours=shift_hours)
		else:
			status = "XX"
		if count:=checkins.get(date):
			if count %2 != 0:
				row["status"] = "ERR"
		temp = {
			"login_date": date,
			"shift": shift_name,
			"status": status
		}
		if not row.get("spent_hours"):
			row["spent_hours"] = None
		temp.update(row)
		result.append(temp)
	return result
	
def get_columns(filters=None):
	columns = [
		{
			"label": _("Login Date"),
			"fieldname": "login_date",
			"fieldtype": "Date"
		},
		{
			"label": _("Shift Name"),
			"fieldname": "shift",
			"fieldtype": "Data"
		},
		{
			"label": _("Status"),
			"fieldname": "status",
			"fieldtype": "Data",
			"width": 120
		},
		{
			"label": _("Late"),
			"fieldname": "late_entry",
			"fieldtype": "Data"
		},
		{
			"label": _("In Time"),
			"fieldname": "in_time",
			"fieldtype": "Data",
			"width":80
		},
		{
			"label": _("Out Time"),
			"fieldname": "out_time",
			"fieldtype": "Data"
		},
		{
			"label": _("Spent Hrs"),
			"fieldname": "spent_hours",
			"fieldtype": "Data"
		},
		{
			"label": _("Late Hrs"),
			"fieldname": "late_hrs",
			"fieldtype": "Data"
		},
		{
			"label": _("Early Hrs"),
			"fieldname": "early_hrs",
			"fieldtype": "Data"
		},
		{
			"label": _("P.Out Hrs"),
			"fieldname": "p_out_hrs",
			"fieldtype": "Data"
		},
		{
			"label": _("Net Wrk Hrs"),
			"fieldname": "net_wrk_hrs",
			"fieldtype": "Data"
		},
		{
			"label": _("OT Hrs"),
			"fieldname": "ot_hours",
			"fieldtype": "Data",
			"width": 120
		},
		{
			"label": _("Total Pay Hrs"),
			"fieldname": "total_pay_hrs",
			"fieldtype": "Data"
		}
	]

	return columns

def get_conditions(filters):
	if not (filters.get("from_date") and filters.get("to_date")):
		frappe.throw("From & To Dates are mandatory")
	conditions = f"""and (at.attendance_date Between '{filters.get("from_date")}' AND '{filters.get("to_date")}')"""
	if filters.get("employee"):
		conditions += f"""and at.employee = '{filters.get("employee")}'"""

	return conditions

def get_date_range(start_date, end_date):
	import datetime
	start_date = getdate(start_date)
	end_date = getdate(end_date)

	range = []
	delta = datetime.timedelta(days=1)
	current_date = start_date

	while current_date <= end_date:
		range.append(current_date)
		current_date += delta

	return range

@frappe.whitelist()
def get_month_range():
	from frappe.utils.dateutils import get_dates_from_timegrain, get_period
	end = today()
	start = add_to_date(end, months=-12)
	periodic_range = get_dates_from_timegrain(start, end, "Monthly")
	periods = [get_period(row) for row in periodic_range]
	periods.reverse()
	return periods

File: gurukrupa_customizations/gurukrupa_customizations/report/monthly_in_out/monthly_in_out.js
// Copyright (c) 2023, 8848 Digital LLP and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Monthly In-Out"] = {
	"filters": [
		{
			"label": __("Month"),
			"fieldtype": "Select",
			"fieldname": "month",
			"reqd": 1,
			"options": [],
			"default": ()=>{
				const dateObject = new Date(); // create a new date object with the current date
				const options = { month: "short", year: "numeric" };
				const dateString = dateObject.toLocaleDateString("en-US", options);
				return dateString
			},
			"on_change": function(query_report){
				var _month = query_report.get_filter_value('month');
				if (!_month) return
				let firstDayOfMonth = moment(_month, "MMM YYYY").toDate();
				firstDayOfMonth = frappe.datetime.obj_to_str(firstDayOfMonth)
				let lastDayOfMonth = frappe.datetime.month_end(firstDayOfMonth)
				query_report.set_filter_value({
					"from_date": firstDayOfMonth,
					"to_date": lastDayOfMonth,
					"employee": null
				});
			}
		},
		{
			"label": __("From Date"),
			"fieldtype": "Date",
			"fieldname": "from_date",
			"read_only": 1,
			"default": frappe.datetime.month_start(frappe.datetime.get_today()),
			"on_change": function(query_report) {
				var from_date = query_report.get_values().from_date;
				if (!from_date) {
					return;
				}
				let date = new moment(from_date)
				var to_date = date.endOf('month').format()
				query_report.set_filter_value({
					"to_date": to_date
				});
			}
		},
		{
			"label": __("To Date"),
			"fieldtype": "Date",
			"fieldname": "to_date",
			"reqd": 1,
			"read_only": 1,
			"default": frappe.datetime.month_end(frappe.datetime.get_today())
		},
		{
			"label": __("Company"),
			"fieldtype": "Link",
			"fieldname": "company",
			"options": "Company",
			"on_change": fetch_employees
		},
		{
			"label": __("Department"),
			"fieldtype": "MultiSelectList",
			"fieldname": "department",
			get_data: function(txt) {
				var company = frappe.query_report.get_filter_value('company');
				var filters = {}
				if (company) filters['company'] = company
				return frappe.db.get_link_options('Department', txt, filters);
			},
			"on_change": fetch_employees
		},
		{
			"label": __("Employees"),
			"fieldname": "employees",
			"fieldtype": "MultiSelectList",
			get_data: function(txt) {
				var filters = get_filter_dict(frappe.query_report)
				if (!filters) filters = {}
				return frappe.db.get_link_options('Employee', txt, filters);
			},
			"on_change": function(query_report){
				var emp_list = query_report.get_filter_value('employees')
				query_report.current_emp = 0
				query_report.emp_count = emp_list.length
				set_employee(query_report, query_report.current_emp)
			},
		},
		{
			"label": __("Selected Employee"),
			"fieldtype": "Link",
			"fieldname": "cur_employee",
			"options": "Employee",
			"reqd": 1,
			"get_query": function() {
				var emp_list = frappe.query_report.get_filter_value('employees')
				return {
					"doctype": "Employee",
					"filters": {"employee":["in",emp_list]}
				}
			},
			"on_change": function(query_report){
				let emp =  frappe.query_report.get_filter_value('cur_employee');
				let employees = query_report.get_filter_value('employees')
				var idx = employees.indexOf(emp)
				if (idx>0) {
					query_report.current_emp = idx
				}
				set_employee_details(query_report)
			},
		},
		{
			"label": __("Employee"),
			"fieldtype": "Link",
			"fieldname": "employee",
			"options": "Employee",
			"reqd": 1,
			"hidden": 1,
			"get_query": function() {
				var filters = get_filter_dict(frappe.query_report)
				if (!filters) return
				return {
					"doctype": "Employee",
					"filters": filters
				}
			}
		},
		{
			"label":"Employee Name",
			"fieldtype": "Data",
			"fieldname": "emp_name",
			"read_only": 1
		},
		{
			"label":"Employee Designation",
			"fieldtype": "Data",
			"fieldname": "designation",
			"read_only": 1
		},
		{
			"label":"Department Name",
			"fieldtype": "Data",
			"fieldname": "emp_department",
			"read_only": 1
		},
		{
			"label":"Punch ID",
			"fieldtype": "Data",
			"fieldname": "punch_id",
			"read_only": 1
		},
		{
			"label":"Shift Hours",
			"fieldtype": "Float",
			"fieldname": "shift_hrs",
			"read_only": 1
		},
		{
			"label": __("Prev"),
			"fieldtype": "Button",
			"fieldname": "prev",
			"depends_on": "eval:(frappe.query_report.emp_count>1 && frappe.query_report.current_emp != 0)",
			"onclick": function() {
				var prev_emp = (frappe.query_report.current_emp - 1)
				set_employee(frappe.query_report, prev_emp, true)
				if (prev_emp < 0) prev_emp = 0
				frappe.query_report.current_emp = prev_emp
			}
		},
		{
			"label": __("Next"),
			"fieldtype": "Button",
			"fieldname": "next",
			"width": "80",
			"depends_on": "eval:(frappe.query_report.emp_count>1 && frappe.query_report.emp_count - 1 > frappe.query_report.current_emp)",
			"onclick": function() {
				var next_emp = (frappe.query_report.current_emp + 1)
				set_employee(frappe.query_report, next_emp, true)
				if (next_emp > frappe.query_report.emp_count) next_emp = frappe.query_report.emp_count - 1
				frappe.query_report.current_emp = next_emp
			}
		},
	],
	"formatter": function(value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);
		var time_columns = ['in_time','out_time','spent_hours','late_hrs','early_hrs','p_out_hrs','net_wrk_hrs','ot_hours','total_pay_hrs']
		if (data && !data.login_date) {
			value = $(`<span>${value}</span>`);
			var $value = $(value).css("font-weight", "bold");
			value = $value.wrap("<p></p>").parent().html();
		}
		else if (in_list(time_columns,column.id) && data.attendance_date) {
			value = frappe.datetime.str_to_user(value, true)
			if (data.late_entry && in_list(["in_time", "out_time"],column.id)) {
				value = $(`<span>${value}</span>`);
				var $value = $(value).css("color", "red");
				value = $value.wrap("<p></p>").parent().html();
			}
			if (column.id == "ot_hours") {
				value = $(`<span>${value}</span>`);
				var $value = $(value).css("color", "green");
				value = $value.wrap("<p></p>").parent().html();
			}
		}
		if (in_list(["ot_hours", "total_pay_hrs"], column.id) && in_list(["Net Hrs w/o OT", "Net Days w/o OT"],data.ot_hours)) {
			value = $(`<span>${value}</span>`);
			var $value = $(value).css("color", "blue");
			value = $value.wrap("<p></p>").parent().html();
		}
		if (column.id == "late_entry" && data.attendance_date) {
			if (data.late_entry){
				value = $(`<input type="checkbox" disabled="" class="disabled-selected">`);
				var $value = $(value);
				value = $value.wrap("<p></p>").parent().html();
			}
			else {
				value = $(`<input type="checkbox" disabled="" class="disabled-deselected">`);
				var $value = $(value);
				value = $value.wrap("<p></p>").parent().html();
			}
		}
		return value;
	},
	onload: (report) => {
		fetch_month_list()
		report.page.add_button("Clear Filters", function() {
			window.open("/app/query-report/Monthly%20In-Out", "_self")
		}).addClass("btn-info")
		report.page.add_button("Generate", function() {
			var emp = frappe.query_report.get_filter_value('cur_employee');
			if (!emp) {
				frappe.msgprint("No Employee Selected")
			}
			else {
				frappe.query_report.set_filter_value({
					"employee": emp
				});
			}
		}).addClass("btn-primary")
		fetch_employees(report)
	}
};

function set_employee(query_report, index, preset = false) {
	var employees = query_report.get_filter_value('employees')
	query_report.set_filter_value({
		"cur_employee": employees[index] || null,
		"employee": preset ? employees[index] : null
	});
}

function set_employee_details(query_report) {
	var employee = query_report.get_filter_value('cur_employee');
	frappe.db.get_value("Employee", employee, ["employee_name", "department", "designation", "attendance_device_id", "default_shift"], (r)=> {
		query_report.set_filter_value({
			"emp_name": r.employee_name,
			"designation": r.designation,
			"emp_department": r.department,
			"punch_id": r.attendance_device_id
		});
		frappe.db.get_value("Shift Type", r.default_shift, "shift_hours", (shift)=>{
			query_report.set_filter_value({
				"shift_hrs": shift.shift_hours
			});
		})
	})
}

function fetch_employees(query_report) {
	var filters = get_filter_dict(query_report)
	if (filters){
		frappe.db.get_list("Employee",{"filters" : filters, "pluck":"name", "order_by":"name"}).then((r)=>{
			query_report.set_filter_value("employees",r)
			if (r.length != 0) {
				query_report.current_emp = 0
				query_report.emp_count = r.length
				set_employee(query_report, query_report.current_emp)
			}
			else {
				query_report.set_filter_value({
					"employee": null
				});		
			}
		})
	}
	else {
		query_report.emp_count = 0
		query_report.set_filter_value({
			"employee": null,
			"employees":[],
			"cur_employee": null,
			"emp_name": null,
			"designation": null,
			"emp_department": null,
			"punch_id": null,
			"shift_hrs": 0
		});
	}
}

function get_filter_dict(query_report) {
	var company = query_report.get_filter_value('company');
	var department = query_report.get_filter_value('department');
	var filters = {}
	if (company && department.length > 0) {
		filters['company'] = company
		filters['department'] = ['in',department]
	}
	else if (company) {
		filters['company'] = company
	}
	else if (department.length > 0) {
		filters['department'] = ['in',department]
	}
	else {
		return null
	}
	return filters
}

function fetch_month_list() {
	frappe.call({
		method: "gurukrupa_customizations.gurukrupa_customizations.report.monthly_in_out.monthly_in_out.get_month_range",
		freeze: false,
		callback: function(r) {
			var month = frappe.query_report.get_filter('month');
				month.df.options.push(...r.message);
				month.refresh();
		}
	})
}